File: segtree.py

class SegmentTree:
    def __init__(self, n: int,defult = 0,func = max) -> None:
        self.tree = [defult]*(2*n)
        self.n = n
        self.func = func
        self.defult = defult


    def make(self,l):
        for i,v in enumerate(l):
            self.tree[self.n + i] = v

        for i in range(self.n-1,0,-1):
            self.tree[i] = self.func(self.tree[2*i],self.tree[2*i+1])

    def sum(self,l = None , r = None) -> int:
        l = (l + self.n) if l is not None else (self.n)
        r = (r + self.n) if r is not None else (2*self.n - 1)
        sm = self.defult
        while l <= r:
            if l&1 == 1:
                sm = self.func(sm,self.tree[l])
                l += 1
            if r&1 == 0:
                sm = self.func(sm,self.tree[r])
                r -= 1
            l >>= 1
            r >>= 1
        return sm

    def add(self,i: int , v) -> None:
        i = i + self.n
        self.tree[i] = v
        i >>= 1
        while i >= 1:
            self.tree[i] = self.func(self.tree[2*i],self.tree[2*i+1])
            i >>= 1


class LazySegTree:
    def __init__(self, n, ID, LID):
        self.n = n
        self.ID = ID      # identity for merge
        self.LID = LID    # identity for lazy
        self.st = [ID] * (4 * n)
        self.lz = [LID] * (4 * n)

    # ===== CUSTOMIZE THESE 3 =====
    def merge(self, a, b):
        return a + b

    def apply(self, p, v, l, r):
        # range add affects sum by (len * v)
        self.st[p] += (r - l + 1) * v

    def compose(self, oldV, newV):
        # range add: add lazies
        return oldV + newV
    # =============================

    def push(self, p, l, r):
        if self.lz[p] == self.LID:
            return
        m = (l + r) // 2
        lc, rc = p * 2, p * 2 + 1

        self.apply(lc, self.lz[p], l, m)
        self.apply(rc, self.lz[p], m + 1, r)

        self.lz[lc] = self.compose(self.lz[lc], self.lz[p])
        self.lz[rc] = self.compose(self.lz[rc], self.lz[p])

        self.lz[p] = self.LID

    def build(self, a, p=1, l=0, r=None):
        if r is None:
            r = self.n - 1
        if l == r:
            self.st[p] = a[l]
            return
        m = (l + r) // 2
        self.build(a, p * 2, l, m)
        self.build(a, p * 2 + 1, m + 1, r)
        self.st[p] = self.merge(self.st[p * 2], self.st[p * 2 + 1])

    def upd(self, ql, qr, v, p=1, l=0, r=None):
        if r is None:
            r = self.n - 1
        if qr < l or r < ql:
            return
        if ql <= l and r <= qr:
            self.apply(p, v, l, r)
            self.lz[p] = self.compose(self.lz[p], v)
            return

        self.push(p, l, r)
        m = (l + r) // 2
        self.upd(ql, qr, v, p * 2, l, m)
        self.upd(ql, qr, v, p * 2 + 1, m + 1, r)
        self.st[p] = self.merge(self.st[p * 2], self.st[p * 2 + 1])

    def qry(self, ql, qr, p=1, l=0, r=None):
        if r is None:
            r = self.n - 1
        if qr < l or r < ql:
            return self.ID
        if ql <= l and r <= qr:
            return self.st[p]

        self.push(p, l, r)
        m = (l + r) // 2
        left = self.qry(ql, qr, p * 2, l, m)
        right = self.qry(ql, qr, p * 2 + 1, m + 1, r)
        return self.merge(left, right)


# ===== Example usage (same as your C++) =====
if __name__ == "__main__":
    n = int(input())
    a = list(map(int, input().split()))

    st = LazySegTree(n, ID=0, LID=0)
    st.build(a)

    st.upd(2, 5, 3)      # add +3 on [2..5]
    print(st.qry(0, 4))  # sum query


class PersistantNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class PersistantSegTree:
    def __init__(self, n):
        self.n = n
        self.T = [PersistantNode()]

    def build(self, arr, node=0, l=0, r=None):
        if r is None:
            r = self.n - 1
        if l == r:
            self.T[node].val = arr[l]
            return
        m = (l + r) // 2
        self.T[node].left = len(self.T)
        self.T.append(PersistantNode())
        self.build(arr, self.T[node].left, l, m)

        self.T[node].right = len(self.T)
        self.T.append(PersistantNode())
        self.build(arr, self.T[node].right, m + 1, r)

        self.T[node].val = self.T[self.T[node].left].val + self.T[self.T[node].right].val
    
    def update(self, idx, val, node=0, l=0, r=None):
        if r is None:
            r = self.n - 1
        if l == r:
            new_node = PersistantNode(val)
            self.T.append(new_node)
            return len(self.T) - 1
        m = (l + r) // 2
        new_node = PersistantNode()
        self.T.append(new_node)
        new_idx = len(self.T) - 1

        if idx <= m:
            self.T[new_idx].left = self.update(idx, val, self.T[node].left, l, m)
            self.T[new_idx].right = self.T[node].right
        else:
            self.T[new_idx].left = self.T[node].left
            self.T[new_idx].right = self.update(idx, val, self.T[node].right, m + 1, r)

        self.T[new_idx].val = self.T[self.T[new_idx].left].val + self.T[self.T[new_idx].right].val
        return new_idx
    
    def query(self, ql, qr, node=0, l=0, r=None):
        if r is None:
            r = self.n - 1
        if qr < l or r < ql:
            return 0
        if ql <= l and r <= qr:
            return self.T[node].val
        m = (l + r) // 2
        left_sum = self.query(ql, qr, self.T[node].left, l, m)
        right_sum = self.query(ql, qr, self.T[node].right, m + 1, r)
        return left_sum + right_sum




if __name__ == '__main__':
    s = SegmentTree(5)
    s.add(0,[4,1])
    s.add(1,[5,3])
    s.add(2,[4,1])
    print(s.sum(0,4))

File: segtree.cpp
#include <bits/stdc++.h>
using namespace std;


template <class Node, class Lazy>
struct LazySegTree {
    int n;
    vector<Node> st;
    vector<Lazy> lz;

    Node ID;
    Lazy LID;

    LazySegTree(int n, Node ID, Lazy LID) : n(n), ID(ID), LID(LID) {
        st.assign(4 * n, ID);
        lz.assign(4 * n, LID);
    }

    Node merge(const Node &a, const Node &b) {
        // merge valuee
        return a + b;
    }

    void apply(int p, const Lazy &v, int l, int r) {
        // apply lazy v to st[p] on segment [l..r]
        st[p] += (r - l + 1) * v;
    }

    Lazy compose(const Lazy &oldV, const Lazy &newV) {
        // how to combine lazies:
        return oldV + newV;
    }

    void push(int p, int l, int r) {
        if (lz[p] == LID) return;
        int m = (l + r) >> 1;
        int lc = p << 1, rc = lc | 1;

        apply(lc, lz[p], l, m);
        apply(rc, lz[p], m + 1, r);

        lz[lc] = compose(lz[lc], lz[p]);
        lz[rc] = compose(lz[rc], lz[p]);

        lz[p] = LID;
    }

    void build(const vector<Node> &a, int p, int l, int r) {
        if (l == r) { st[p] = a[l]; return; }
        int m = (l + r) >> 1;

        build(a, p << 1, l, m);
        build(a, p << 1 | 1, m + 1, r);

        st[p] = merge(st[p << 1], st[p << 1 | 1]);
    }

    void build(const vector<Node> &a) { build(a, 1, 0, n - 1); }

    void upd(int ql, int qr, const Lazy &v, int p, int l, int r) {
        if (qr < l || r < ql) return;
        if (ql <= l && r <= qr) {
            apply(p, v, l, r);
            lz[p] = compose(lz[p], v);
            return;
        }
        push(p, l, r);
        int m = (l + r) >> 1;
        upd(ql, qr, v, p << 1, l, m);
        upd(ql, qr, v, p << 1 | 1, m + 1, r);
        st[p] = merge(st[p << 1], st[p << 1 | 1]);
    }

    void upd(int l, int r, const Lazy &v) { upd(l, r, v, 1, 0, n - 1); }

    Node qry(int ql, int qr, int p, int l, int r) {
        if (qr < l || r < ql) return ID;
        if (ql <= l && r <= qr) return st[p];
        push(p, l, r);
        int m = (l + r) >> 1;
        return merge(qry(ql, qr, p << 1, l, m),
                     qry(ql, qr, p << 1 | 1, m + 1, r));
    }

    Node qry(int l, int r) { return qry(l, r, 1, 0, n - 1); }
};

using Node = long long;
using Lazy = long long;
struct Seg : LazySegTree<Node, Lazy> {
    Seg(int n) : LazySegTree(n, 0LL, 0LL) {}

    Node merge(const Node &a, const Node &b) {
        return a + b;
    }
    void apply(int p, const Lazy &v, int l, int r) {
        st[p] += (r - l + 1) * v;
    }
    Lazy compose(const Lazy &oldV, const Lazy &newV) {
        return oldV + newV;
    }
};

int main () {


    int n; cin >> n;
    vector<Node> a(n);
    for (auto &x : a) cin >> x;



    Seg st(n);
    st.build(a);

    st.upd(2, 5, 3);         // add +3 on [2..5]
    cout << st.qry(0, 4);    // sum query

}